File: days/day_05/solutions/solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

class Solution {
    public static void main(String[] args) throws IOException {
        Path inputPath = Paths.get("../data/input.txt");
        String content = new String(Files.readAllBytes(inputPath));
        long[] results = solve(content);
        System.out.println("Part 1: " + results[0]);
        System.out.println("Part 2: " + results[1]);
    }
    
    private static long[] solve(String inputData) {
        String[] lines = inputData.trim().split("\n");
        
        // Find blank line separator
        int blankIdx = -1;
        for (int i = 0; i < lines.length; i++) {
            if (lines[i].trim().isEmpty()) {
                blankIdx = i;
                break;
            }
        }
        
        // Parse ranges (first section)
        List<long[]> ranges = new ArrayList<>();
        for (int i = 0; i < blankIdx; i++) {
            String line = lines[i].trim();
            if (line.isEmpty()) continue;
            String[] parts = line.split("-");
            long start = Long.parseLong(parts[0]);
            long end = Long.parseLong(parts[1]);
            ranges.add(new long[]{start, end});
        }
        
        // Parse IDs to check (second section)
        List<Long> ids = new ArrayList<>();
        for (int i = blankIdx + 1; i < lines.length; i++) {
            String line = lines[i].trim();
            if (line.isEmpty()) continue;
            ids.add(Long.parseLong(line));
        }
        
        // Part 1: Count how many IDs fall into any range
        int part1Count = 0;
        for (long idVal : ids) {
            for (long[] range : ranges) {
                if (range[0] <= idVal && idVal <= range[1]) {
                    part1Count++;
                    break;
                }
            }
        }
        
        // Part 2: Merge ranges and count total unique IDs covered
        // Sort ranges by start value
        ranges.sort((a, b) -> Long.compare(a[0], b[0]));
        
        // Merge overlapping/adjacent ranges
        List<long[]> merged = new ArrayList<>();
        if (!ranges.isEmpty()) {
            merged.add(new long[]{ranges.get(0)[0], ranges.get(0)[1]});
            for (int i = 1; i < ranges.size(); i++) {
                long[] current = ranges.get(i);
                long[] last = merged.get(merged.size() - 1);
                // Check if overlaps or is adjacent (current[0] <= last[1] + 1)
                if (current[0] <= last[1] + 1) {
                    // Merge: update end to max of both ends
                    last[1] = Math.max(last[1], current[1]);
                } else {
                    // No overlap, add as new range
                    merged.add(new long[]{current[0], current[1]});
                }
            }
        }
        
        // Calculate total unique IDs covered
        long part2Total = 0;
        for (long[] range : merged) {
            part2Total += range[1] - range[0] + 1;
        }
        
        return new long[]{part1Count, part2Total};
    }
}

File: days/day_05/solutions/solution.ex
defmodule Solution do
  def solve(input_data) do
    lines = String.split(String.trim(input_data), "\n")

    # Find blank line separator
    blank_idx = Enum.find_index(lines, fn line -> String.trim(line) == "" end) || length(lines)

    # Parse ranges (first section)
    ranges = lines
      |> Enum.take(blank_idx)
      |> Enum.filter(&(String.trim(&1) != ""))
      |> Enum.map(fn line ->
        [start_str, end_str] = String.split(line, "-")
        {String.to_integer(start_str), String.to_integer(end_str)}
      end)

    # Parse IDs to check (second section)
    ids = lines
      |> Enum.drop(blank_idx + 1)
      |> Enum.filter(&(String.trim(&1) != ""))
      |> Enum.map(&String.to_integer/1)

    # Part 1: Count how many IDs fall into any range
    part1_count = Enum.count(ids, fn id_val ->
      Enum.any?(ranges, fn {start, e} -> start <= id_val && id_val <= e end)
    end)

    # Part 2: Merge ranges and count total unique IDs covered
    # Sort ranges by start value
    sorted_ranges = Enum.sort_by(ranges, fn {start, _} -> start end)

    # Merge overlapping/adjacent ranges
    merged = if sorted_ranges == [] do
      []
    else
      {merged_ranges, _} = Enum.reduce(sorted_ranges, {[], nil}, fn {curr_start, curr_end}, {acc, last} ->
        if last == nil do
          {[[curr_start, curr_end]], {curr_start, curr_end}}
        else
          {last_start, last_end} = last
          if curr_start <= last_end + 1 do
            # Merge: update end to max of both ends
            new_last = {last_start, max(last_end, curr_end)}
            new_acc = List.replace_at(acc, length(acc) - 1, [last_start, max(last_end, curr_end)])
            {new_acc, new_last}
          else
            # No overlap, add as new range
            {acc ++ [[curr_start, curr_end]], {curr_start, curr_end}}
          end
        end
      end)
      merged_ranges
    end

    # Calculate total unique IDs covered
    part2_total = Enum.reduce(merged, 0, fn [start, e], sum -> sum + (e - start + 1) end)

    {part1_count, part2_total}
  end
end

content = File.read!("../data/input.txt")
{part1, part2} = Solution.solve(content)
IO.puts("Part 1: #{part1}")
IO.puts("Part 2: #{part2}")
